Use poll in readBytes to fix segfaults with high fd numbers

hiddenalpha opened the pull request:
- Replace 'select' by 'poll' in readBytes to fix segfaults when fd >= 1024
- Log a few error cases which we concealed in previous code.
- Fix some concurrency issues.

HINT: I did not add recompiled libs as I only could provide/test the linux lib
anyway. Not sure how this gets handled in this git repo. Let me know.

**Edit: (@tresf)** 
* Supersedes #78
* Closes #53

Commit:
Use poll to fix segfaults with high fd numbers

- Replace 'select' by 'poll' to fix segfaults when fd >= 1024
- Log a few error cases which we concealed in old code.
- Fix some concurrency issues.

(wip 4a6ff6cfe89db16040c95dc59fb5d7efb9639a46)

tresf commented on src/main/cpp/_nix_based/jssc.cpp:
```diff
@@ -49,6 +50,13 @@
+    #define DEBUG 0
+#endif
+#define LOG_WARN( fmt, ... ) fprintf(stderr, "[WRN %s:%d] " fmt, strrchr(__FILE__,'/')+1, __LINE__,  __VA_ARGS__ )
+#define LOG_DEBUG( fmt, ... ) if(DEBUG)fprintf(stderr, "[DBG %s:%d] " fmt, strrchr(__FILE__,'/')+1, __LINE__, __VA_ARGS__ );
```
Is it normal to write to `stderr` for debug?

hiddenalpha replied:
I think this is the same question as [stackoverflow.com/q/4919093/4415884](https://stackoverflow.com/q/4919093/4415884).

**PRO**: Logging to stdout would make it hard (if not impossible) to write a program which produces data to stdout (Eg somewhat like `git archive > my.tar`) or also with pipes (Eg `tar cz * | ssh example.com -- tar xz`). If any library writes to stdout this will no longer be possible as it would "destroy" the actual data.

**CON**: AFAIK stderr is unbuffered. I never had an issue with this. But that does not mean it could have an impact which I'm not aware of.

**Finally**: In "my project" it does not matter as we do not use stdout for data. So "my project" would be fine with logging to stdout too. But I don't know about others. That's why I added the idea with slf4j. So the user could decide himself.

WDYT?

tresf commented on src/main/cpp/_nix_based/jssc.cpp:
```diff
@@ -527,27 +535,70 @@ JNIEXPORT jboolean JNICALL Java_jssc_SerialNativeInterface_writeBytes
 JNIEXPORT jbyteArray JNICALL Java_jssc_SerialNativeInterface_readBytes
-  (JNIEnv *env, jobject object, jlong portHandle, jint byteCount){
-    fd_set read_fd_set;
+  (JNIEnv *env, jobject thisObj, jlong portHandle, jint byteCount){
+
+    // Error-Handling:
+    // I have no idea why we just ignored any kind of errors (eg ignoring return
+    // values of 'select' and 'read') in the old code. I also could not find any
+    // comments which would explain why. I liked to communicate those by raising
+    // java exceptions. But this would not be backward compatible. So I restrict
+    // myself to just log the irregular cases for now. If I get some time, I might
+    // open another PR to add some error-handling. So we could discuss that topic
+    // there.
```
You can comment on your own PR lines to explain things like these (or open as new issues).  The code should only include the appropriate `TODO`'s.

tresf replied:
```suggestion
  (JNIEnv *env, jobject thisObj, jlong portHandle, jint byteCount){

    // TODO: Errors should be communicated by raising java exceptions (will break backwards compat)
```

tresf commented on src/main/cpp/_nix_based/jssc.cpp:
```diff
@@ -49,6 +50,13 @@
+#ifndef DEBUG
+    #define DEBUG 0
+#endif
+#define LOG_WARN( fmt, ... ) fprintf(stderr, "[WRN %s:%d] " fmt, strrchr(__FILE__,'/')+1, __LINE__,  __VA_ARGS__ )
```
I've never seen the words "WARN" and "DEBUG" abbreviated.  I'm interested to know how standard this is.   Also, any debug capabilities should be consistent across platforms, so this should also go into the win32 stuff.

hiddenalpha replied:
Just a silly idea of mine. I think you're right and we should write them out 👍

tresf commented on src/main/cpp/_nix_based/jssc.cpp:
```diff
@@ -49,6 +50,13 @@
 
 //#include <iostream> //-lCstd use for Solaris linker
 
+// TODO: IMHO logging to slf4j would be nice.
```
```suggestion
// TODO: Move logging to java
```

tresf commented:
Thanks.  I'd also ask you to read the code and suggestions here: https://github.com/java-native/jssc/pull/78.  Note, we're short on volunteers here to help merge, so I do apologize for any delay before this hits our master branch.

hiddenalpha commented:
Thx for the hint 👍 Seems as we need a few more tweaks then. Will give it a try.

Commit:
Apply suggested changes (3rd)

- Reword some comments.
- Drop logging.
- Use 'select' as fallbck for OsX. But still allow to override using
  -DHAVE_POLL=1 gcc switch.
- Revert accidentally checked-in libjssc.so.

(wip 7e4e60e0e5264064c6bf069c34b5d27e5c8131cd)

Commit:
Merge upstream changes from master to feature

(wip 96ab642a3b1888fa5411f60d85058dd37b742160)

graynk reviewed (commented):
Is this part still needed? It seems like in this implementation we wait for data before returning in native code (either by poll or by select, even on Mac OS), so this sleep can be removed, if I understand everything correctly

https://github.com/java-native/jssc/blob/597d06afd75b8b479d8d2ec57d7bc807eee186ba/src/main/java/jssc/SerialPort.java#L1304-L1310

graynk commented on src/main/cpp/_nix_based/jssc.cpp:
```diff
@@ -524,28 +534,90 @@ JNIEXPORT jboolean JNICALL Java_jssc_SerialNativeInterface_writeBytes
+    while(true){
+        int result = poll(fds, 1, -1);
+        if(result < 0){
+            // man poll: On error, -1 is returned, and errno is set to indicate the error.
```
won't error be returned in case the port has been closed? what would happen if we call `readBytes` and then `closePort`? won't this just continue indefinitely?

hiddenalpha replied:
AFAIK this ends up in an infinite-loop exactly as the currently released code does (by [ignoring return code of *select()*](https://github.com/java-native/jssc/blob/v2.9.2/src/main/cpp/_nix_based/jssc.cpp#L540)). So its at least "backward compatible" ;)

Further there's also an infinite-loop in case of EOF because code currently just loops if "*read() < 0*" (See [code of v2.9.2](https://github.com/java-native/jssc/blob/v2.9.2/src/main/cpp/_nix_based/jssc.cpp#L542)). (It's easily reproducible: Just pull the usb cable while reading and observe what happens with the CPU)

I feel there is a lot of error-handling missing everywhere in native code. But I decided against trying to fix everything in a single PR at once because:
- Tends to end up in messy (as in *hard-to-review*) PRs.
- Will bring breaking changes in most cases.

Maybe I should open individual issues to document all my findings (currently only visible here and there in my scratch-book).

WDYT?

tresf replied:
I'm hoping for a palatable solution. 

> this ends up in an infinite-loop exactly as the currently released code does

Should we just merge as-is and handle the exceptions at a later time?

tresf replied:
Checking in to see if we're in agreement on merging this?

graynk replied:
Sure, I don't mind. There's still `Thread.sleep` mentioned before this, which I don't think is needed anymore though. But I'm fine merging as-is

tresf replied:
> there's still `Thread.sleep` mentioned before this, which I don't think is needed anymore though. But I'm fine merging as-is

It was factored out but GitHub for some reason didn't put the `outdated` keyword on the message.  It also doesn't allow me to reply or resolve it for some reason.  From what I can see, it's removed from the PR.

tresf replied:
Whoops, it's still there.  https://github.com/java-native/jssc/blob/1e9b04ebe53c0724245def5bd10d92a190aeeebf/src/main/java/jssc/SerialPort.java#L559.  @hiddenalpha, thoughts?

GMKennedy replied:
Sleep has been disabled, no issues encountered as far as higher CPU loading after applying monitor lock to act in its place globally in SerialPort.java

tresf replied:
Sleep removed per 1aff5f9.

tresf commented:
> Maybe I should open individual issues to document all my findings (currently only visible here and there in my scratch-book) so we could discuss what we want do about them.

Fine by me.  @GMKennedy and I will be making a new release this week, so I'd like to get this merged in asap if possible.  For the record @GMKennedy did compile and test the code on MacOS and Linux with no observable regressions.

tresf commented:
Just checking in on the `Sleep(...)` comments.  I'd like to have a decision on this so we can make this part of the next release.

GMKennedy reviewed (approved):
Code changes implement 'synchronized' function (ie monitor lock) which replaces thread sleep function for handling of contention/concurrency.  SerialPort.java has been updated accordingly.

tresf commented:
AFAIK, @GMKennedy's speaking specifically about this branch: https://github.com/java-native/jssc/commits/Poll-Update

We'll add to this PR  in the next day and merge.

tresf committed:
Remove sleep()

hiddenalpha commented:
> @hiddenalpha, Just checking in on the `Sleep(...)` comments. I'd like to have a decision on this so we can make this part of the next release.

Sorry for not responding. But had no access to my account because I was not at work a few weeks. AFAIK this decision is superfluous meanwhile :) But honestly I even could not say if it should be removed or not.

factoritbv commented:
During our thorough testing (mac/linux/raspberry) of our improvements suggested in #126, we have never run into the issue of file descriptors with a value above 1024. However, it seems to be an issue on (some?) linux systems.

Maybe you could have a look and improve our suggested code with your polling mechanism. I'm not sure if it is as portable as `select()`, which is basically available on every UNIX system. But if it makes it more stable / flexible, I think the code in #126 can be improved with use of the `poll()` function.

hiddenalpha commented:
> ... I'm not sure if it is as portable as `select()`, ...

Yeah :( unfortunately poll is not as portable as select. This was the reason we kept select as a fallback.

> ... we have never run into the issue of file descriptors with a value above 1024. However, it seems to be an issue on (some?) linux systems.

Unluckily I do not know the exact reason why we ran into such FDs. I only saw that we had them. There were some suspicions that it could come for example from [docker](https://www.docker.com/) or that we use the lib in a jetty where half a dozen apps are running performing lots of file/network IO.
